File: api/lib/services/caseService.ts
import { supabaseAdmin } from "../supabase";
import type { Case, TablesInsert } from "../types";

export async function listCases(userId: string): Promise<Case[]> {
  const { data, error } = await supabaseAdmin
    .from("cases")
    .select("*")
    .eq("user_id", userId)
    .order("updated_at", { ascending: false });
  if (error) {
    throw error;
  }
  return data as Case[];
}

export async function createCase(payload: TablesInsert["cases"]): Promise<Case> {
  const { data, error } = await supabaseAdmin.from("cases").insert(payload).select().single();
  if (error) {
    throw error;
  }
  return data as Case;
}

export async function getCase(caseId: string): Promise<Case | null> {
  const { data, error } = await supabaseAdmin.from("cases").select("*").eq("id", caseId).maybeSingle();
  if (error) {
    throw error;
  }
  return (data as Case) ?? null;
}

export async function updateCase(caseId: string, updates: Partial<TablesInsert["cases"]>): Promise<Case | null> {
  const { data, error } = await supabaseAdmin
    .from("cases")
    .update(updates)
    .eq("id", caseId)
    .select()
    .maybeSingle();
  if (error) {
    throw error;
  }
  return (data as Case) ?? null;
}

export async function deleteCase(caseId: string): Promise<void> {
  const { error } = await supabaseAdmin.from("cases").delete().eq("id", caseId);
  if (error) {
    throw error;
  }
}

File: backend/app/services/case_service.py
from __future__ import annotations

from typing import Any

from ..models import Case, CaseCreate, CaseUpdate
from .supabase_service import SupabaseService


class CaseService:
    def __init__(self, supabase: SupabaseService | None = None):
        self.supabase = supabase or SupabaseService()

    def list_cases(self, user_id: str) -> list[Case]:
        records = self.supabase.select("cases", filters={"user_id": user_id}, order=("updated_at", False))
        return [Case(**record) for record in records]

    def create_case(self, user_id: str, payload: CaseCreate) -> Case:
        data: dict[str, Any] = payload.model_dump()
        insert_payload = data | {"user_id": user_id}
        record = self.supabase.insert("cases", insert_payload)[0]
        return Case(**record)

    def get_case(self, case_id: str) -> Case | None:
        record = self.supabase.select_by_id("cases", case_id)
        return Case(**record) if record else None

    def update_case(self, case_id: str, payload: CaseUpdate) -> Case | None:
        data = {k: v for k, v in payload.model_dump(exclude_none=True).items()}
        if not data:
            return self.get_case(case_id)
        record = self.supabase.update("cases", case_id, data)[0]
        return Case(**record)

    def delete_case(self, case_id: str) -> None:
        self.supabase.delete("cases", case_id)

